### PR TITLE
Convert pcl_conversions to ROS2.

### PR DESCRIPTION
**Pull request by clalancette**
This is a pretty straightforward conversion.  The only slightly interesting bit is that I ended up disabling all of the parts having to do with pcl_msgs, since we don't yet have that in ROS2.

connects to ros2/ros2#342

### CODE REVIEW

File: CMakeLists.txt
Comment (mikaelarguedas) [inline CMakeLists.txt]: please use c++14 to stay consistent with the rest of the codebase

File: package.xml
Comment (mikaelarguedas) [inline package.xml:31]: I think this is needed at runtime too (exec_depend), aren't they?

File: test/test_pcl_conversions.cpp
Comment (mikaelarguedas) [inline test/test_pcl_conversions.cpp]: Do you need this debug print ? if yes can you print to stdout given that it's not an error ?

File: include/pcl_conversions/pcl_conversions.h
Comment (mikaelarguedas) [inline include/pcl_conversions/pcl_conversions.h]: Nit: please use space after comments

File: include/pcl_conversions/pcl_conversions.h
Comment (mikaelarguedas) [inline include/pcl_conversions/pcl_conversions.h]: Do we need to comment out all these includes ?

File: include/pcl_conversions/pcl_conversions.h
Comment (mikaelarguedas) [inline include/pcl_conversions/pcl_conversions.h:88]: can we keep the comment as is ? (from ns)

File: include/pcl_conversions/pcl_conversions.h
Comment (mikaelarguedas) [inline include/pcl_conversions/pcl_conversions.h:81]: cf https://github.com/ros2/cartographer_ros/pull/1/commits/3c3c2c277f47c1c4da153f086f1a2f70a5619283

**Review by mikaelarguedas — COMMENTED** [review]
Thanks for pounding on that! Lookg good, a few inline comments / questions before approving

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt]: Will do.

File: package.xml
Comment (clalancette) [reply package.xml:31]: So, I'm not clear about the exact semantics of exec_depend with a header-only library such as this one.  Technically, it has zero runtime requirements, since it is header-only.  But I did notice that the ROS1 version of this package has <run_depend>, so I'm not sure.  Any advice?

File: test/test_pcl_conversions.cpp
Comment (clalancette) [reply test/test_pcl_conversions.cpp]: No, left that in by accident.  I'll remove it.

File: include/pcl_conversions/pcl_conversions.h
Comment (clalancette) [reply include/pcl_conversions/pcl_conversions.h]: Technically, I can leave in the pcl/ ones.  Those should work.  They just aren't necessary/useful until we uncomment the pcl_msgs versions, and I think it is a little cleaner since the TODO comment applies to the whole block.  That being said, I don't care all that much, so if you think I should uncomment them, I will.  Just let me know.

File: include/pcl_conversions/pcl_conversions.h
Comment (clalancette) [reply include/pcl_conversions/pcl_conversions.h]: Will fix.

File: include/pcl_conversions/pcl_conversions.h
Comment (clalancette) [reply include/pcl_conversions/pcl_conversions.h:81]: Yeah, I should just use that.  I'll switch it.

File: package.xml
Comment (mikaelarguedas) [reply package.xml:31]: oh in that case that's fine as is, sorry for the noise

File: include/pcl_conversions/pcl_conversions.h
Comment (mikaelarguedas) [reply include/pcl_conversions/pcl_conversions.h]: That's fine, porting `pcl_msgs` should be very straight forward so we can reenable them all at once :+1:

File: include/pcl_conversions/pcl_conversions.h
Comment (clalancette) [reply include/pcl_conversions/pcl_conversions.h:88]: So, I changed it because technically we aren't converting from nanoseconds anymore; we are converting from sec/nsec together.  It's totally clear what we are converting *from* based on the code; I thought the comment might be helpful for what we are converting *to*.  However, I wonder if it might just be better to change "pcl_stamp" to "pcl_stamp_us", and just get rid of the comment completely.  What do you think?

File: CMakeLists.txt
Comment (mikaelarguedas) [inline CMakeLists.txt]: that does look weird, which one prevails ? QUIET or REQUIRED ? (I know this hasn't been modified in this PR, just curious)

Comment (clalancette) [reply CMakeLists.txt]: Heh, good question, that is weird.  Reading: https://cmake.org/pipermail/cmake/2007-June/014834.html, it seems like those keywords set an XXX_FIND_REQUIRED and XXX_FIND_QUIET variable before invoking the find.  So I guess it just depends on which order the "if" statements are in the code.  @wjwwood You actually put this in initially; what's the semantics and/or reasoning behind this?

Comment (wjwwood) [reply CMakeLists.txt]: Quiet just means it will suppress extra console output, it does not prevent it from failing. Required just makes it fail if it is not found. These options are sort of orthogonal.

File: include/pcl_conversions/pcl_conversions.h
Comment (clalancette) [reply include/pcl_conversions/pcl_conversions.h:81]: Actually, it turns out that there is a good reason *not* to use the scientific notation.  When you specify something like "1e9", it is implicitly a double.  In the cases above, I actually have to have the unsigned long (otherwise I get compilation errors).  So either we use the long form 1000000ull (yuck) or we use the scientific notation with a cast (unsigned long long)1e9 (double yuck).  I opted to stay with the original long form, but let me know if you think differently.

**Comment by clalancette** [comment]
windows: [Build Status](http://ci.ros2.org/job/ci_windows/2592/)
linux: [Build Status](http://ci.ros2.org/job/ci_linux/2472/)
osx: [Build Status](http://ci.ros2.org/job/ci_osx/1928/)

File: include/pcl_conversions/pcl_conversions.h
Comment (mikaelarguedas) [reply include/pcl_conversions/pcl_conversions.h:81]: Sorry for the oversight, :+1: for keeping it as is

**Comment by clalancette** [comment]
This failed on OSX and Windows because they both lack libpcl in the build farm.

It should be pretty straightforward to fix the OSX problem, just by installing libpcl from Homebrew.  I've opened up an issue for that here: https://github.com/ros2/build_cop/issues/6

Windows is probably going to be more interesting.  I've opened up an issue for that here: https://github.com/ros2/choco-packages/issues/1

Note, however, that if we restrict pcl_conversions to the turtlebot demo, we don't have to wait for either of those to be resolved.

**Comment by nuclearsandwich** [comment]
A rebuild failed with the same CMake error. Neither the homebrew libpcl package nor CMake provide a CMake package config or FindPCL.cmake. It looks like we get the same build error on Linux which may mean libpcl is missing there too or it may be another missing module/config situation. Though I thought cmake had a pkgconfig fallback it could use. :dizzy:

**Comment by clalancette** [comment]
I swear the Linux one passed yesterday.  I may have been doing too many things at once.

This definitely works locally, where I have libpcl-dev installed on my machine.  I'll take another look at it and see if I can figure out where the cmake files are coming from.

**Comment by nuclearsandwich** [comment]
Some of the info here may or may not be helpful: https://cmake.org/Wiki/CMake:How_To_Find_Libraries#Piggybacking_on_pkg-config

**Comment by clalancette** [comment]
Due to the dependency on libpcl (which we haven't figured out on OSX or Windows), this is going to be a Linux-only, turtlebot-demo-only conversion for now.  CI for just Linux is below:

linux: [Build Status](http://ci.ros2.org/job/ci_turtlebot-demo/13/)

**Comment by clalancette** [comment]
@ros2/team I think this one is ready to go, just needs review/approval.

File: CMakeLists.txt
Comment (mikaelarguedas) [inline CMakeLists.txt:10]: If you find_package Boost you need to add it to the dependencies in the package.xml.

In this case you don't use boost anywhere (the only occurences of boost are for shared_ptr and commented out) so I'd suggest removing the find_package and convert them to std::shared_ptr

File: package.xml
Comment (mikaelarguedas) [inline package.xml:2]: please do package conversion to format 2 in a separate commit so that it can be submitted upstream easily. https://github.com/ros-perception/pcl_conversions/issues/26#issue-224155516

Comment (clalancette) [reply package.xml:2]: Will do.

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt:10]: Unfortunately, libpcl itself uses boost all over the place.  Thus, we need to have boost both in CMakeLists.txt and in package.xml, since it truly is a dependency here (you cannot successfully link without boost).

**Comment by clalancette** [comment]
I've redone these commits, splitting out the conversion to package format 2 (which I'll open a PR for upstream).  Other than that, here's the CI for it:

linux: [Build Status](http://ci.ros2.org/job/ci_turtlebot-demo_linux/8/)

**Comment by clalancette** [comment]
@Karsten1987 If you could review/merge this, that would be helpful.  Thanks!

**Comment by mikaelarguedas** [comment]
Don't have time to review this but it would be good to discuss what we do about `seq` before merging this

**Comment by clalancette** [comment]
Sorry, I don't quite understand.  What do you mean by `seq`?

**Comment by mikaelarguedas** [comment]
I was referring to the TODOs added in this PR
```
// TODO(clalancette): Seq doesn't exist in the ROS2 header
// anymore.  Do we need to do anything here?```

**Comment by clalancette** [comment]
Haha.  I had forgotten about that bit.  I was searching for seq and couldn't find it, but yeah, we should figure out what to do about it.

**Comment by mikaelarguedas** [comment]
```
// TODO(clalancette): Seq doesn't exist in the ROS2 header
// anymore.  Do we need to do anything here?
```
@ros2/team following up on this, should we just get rid of any reference to the sequence number field of the header message here ?

**Comment by wjwwood** [comment]
@mikaelarguedas can you do a line comment? your snippet doesn't have any context.

File: include/pcl_conversions/pcl_conversions.h
Comment (mikaelarguedas) [inline include/pcl_conversions/pcl_conversions.h]: @ros2/team should we just get rid of any reference to the sequence number field of the header message here (and below)?

**Comment by mikaelarguedas** [comment]
> @mikaelarguedas can you do a line comment? your snippet doesn't have any context.

Done https://github.com/ros2/pcl_conversions/pull/1#discussion_r119474963, sorry about that

File: include/pcl_conversions/pcl_conversions.h
Comment (wjwwood) [reply include/pcl_conversions/pcl_conversions.h]: Yes, I think you can safely just get rid of it.

File: include/pcl_conversions/pcl_conversions.h
Comment (wjwwood) [inline include/pcl_conversions/pcl_conversions.h]: There should be some way to get the sequence number from the system in the future (part of the rationale was that the middleware should be able to answer this question for all messages anyways and either per publisher or per received), but for now leaving it 0 is fine by me.

**Comment by wjwwood** [comment]
> Done #1 (comment), sorry about that

No worries, I just couldn't find it, it turns out the file view was not loading the full diff because it was big so ctrl-f didn't find the TODO.

**Comment by clalancette** [comment]
I cleaned up the comments around seq a bit, to add the clarification from @wjwwood .  Otherwise, I think we're good to go on this PR.  Anyone want to approve?

**Review by wjwwood — APPROVED** [review]

File: include/pcl_conversions/pcl_conversions.h
Comment (mikaelarguedas) [inline include/pcl_conversions/pcl_conversions.h]: :+1: for the new comment. This line can be removed though given that header will not have a seq field in the future

**Comment by mikaelarguedas** [comment]
@clalancette please update maintainer tags when porting packages

Note: actually some other packages didn't update it. Maybe it's ok to leave them as is as long as we dont send emails for failing builds on the ROS2 buildfarm ?

In any case :+1: for merging

**Comment by nuclearsandwich** [comment]
> Maybe it's ok to leave them as is as long as we dont send emails for failing builds on the ROS2 buildfarm ?

At the moment, email is completely unwired on the ROS 2 test farm. When we switch to the prime farm I was not planning to send emails to anyone but myself and other volunteers for beta 2.

File: include/pcl_conversions/pcl_conversions.h
Comment (clalancette) [reply include/pcl_conversions/pcl_conversions.h]: Yeah, good point.  I've killed the line of code now.

**Comment by clalancette** [comment]
And I ended up fixing the maintainer tag.  I'm going to merge now, thanks!

**Pull request merged**
